packages/commonwealth/client/scripts/views/pages/contracts/add_contract_form.tsx

```diff
@@ -0,0 +1,180 @@
+/* @jsx m */
+
+import m from 'mithril';
+import $ from 'jquery';
+
+import 'pages/create_community.scss';
+
+import app from 'state';
+import {
+  ChainBase,
+  ChainNetwork,
+  ChainType,
+  ContractType,
+} from 'common-common/src/types';
+import { isAddress } from 'web3-utils';
+
+import { notifyError } from 'controllers/app/notifications';
+import { IdRow, InputRow, SelectRow } from 'views/components/metadata_rows';
+
+import { CWButton } from 'views/components/component_kit/cw_button';
+import { CWValidationText } from 'views/components/component_kit/cw_validation_text';
+
+import { linkExistingAddressToChainOrCommunity } from 'controllers/app/login';
+
+import {
+  initChainForm,
+  defaultChainRows,
+  ethChainRows,
+} from '../create_community/chain_input_rows';
+
+import {
+  ChainFormFields,
+  ChainFormState,
+  EthChainAttrs,
+  EthFormFields,
+} from '../create_community/types';
+
+type ContractFormFields = {
+  chain_node_id: number;
+  abi: JSON;
+  contractType: ContractType;
+  decimals: number;
+  token_name: string;
+};
+
+type CreateContractForm = ChainFormFields & EthFormFields & ContractFormFields;
+
+type CreateContractState = ChainFormState & { form: CreateContractForm };
+
+export class AddContractForm implements m.ClassComponent<EthChainAttrs> {
+  private state: CreateContractState = {
+    message: '',
+    loaded: false,
+    loading: false,
+    saving: false,
+    status: undefined,
+    form: {
+      address: '',
+      chainString: 'Ethereum Mainnet',
+      // For Now we hard code the chain node id until we have a better way to distinguish between chains
+      chain_node_id: 37,
```

Comment:
See how `ethChainRows` does it -- is that useful? Why do we even need the database nodeId at all?

Reply:
we need the nodeId because the contract exists on a chain?

Reply:
yeah, it was originally done the way ethChainRows does it but Zak and I wanted to make it able to be chain agnostic down the line